src/PostgREST/Plan.hs

```diff
@@ -431,26 +553,27 @@ mutatePlan mutation qi ApiRequest{..} sCache readReq = mapLeft ApiRequestError $
           Left InvalidFilters
     MutationDelete -> Right $ Delete qi combinedLogic iTopLevelRange rootOrder returnings
   where
+    ctx = ResolverContext dbTables dbRepresentations qi "json"
```

Comment:
The `"json"` here, perhaps it could be a type instead of `Text`?

Also, isn't this `"json"` because the payload comes with `Content-Type: application/json`? Maybe we could reuse `MTApplicationJSON` here somehow.

Reply:
> The "json" here, perhaps it could be a type instead of Text?

Yes I think it could be, although...

> Also, isn't this "json" because the payload comes with Content-Type: application/json?

Yes and no, in **this** particular line it's JSON because that's what we want the **response** format to be (`outputType` in `ResolverContext`). So it chooses which data rep(s) to use when we're outputting for `return=representation`. However, the type for the input payload is also hardcoded down below when `typedColumnsOrError` uses `withJsonParse` internally. So you're absolutely right.

It could be weird or possibly even wrong to hardcode it as `json`, but my thinking was that we don't have data rep support for any other output format so might as well format it as JSON rather than not formatting it at all. Like even in say CSV output (is that possible for `return=representation`? I never tried it), in the absense of full data reps support, the `domain->json` function is probably better than outputting the raw underlying type (which could be binary).

What do you think?

Reply:
Still thinking about this after you mentioned `MTApplicationJSON`. 

I'm not sure how this feature _should_ play out when the requested content type (or the payload content type) isn't JSON.

Look at `bytea` data. When trying to get the raw binary data for a `bytea` column with `Accept: application/octet-stream`... I think today data reps are bypassed entirely when you make a raw request via the `binaryField` bits. The binary request code doesn't care what the column type is. That's good in a way, that's most likely exactly what you wanted as a user. On the other hand it's technically incorrect, you requested to receive a column containing `base64` (as defined by your data reps) so you should receive not the `bytea` binary but the utf-8 encoded base64 encoded `bytea` data.

One could also imagine data reps being used as a security thing. Dumb example but imagine a data rep that lets you write your credit card number into a table but when you try to read it back it only gives you the last 4 digits. So you can bypass that whole mechanism by requesting binary instead. (Not that I think we should be making any kind of security guarantees with data reps, it's a convenience feature to make a more expressive API. But still, it might be surprising that this is possible. Surprising is usually bad.)

So what to do? If we respect the data rep for single field octet stream, you can't get the underlying binary data easily without making another view or an RPC function. Maybe that's okay. Or we could allow an "escape hatch" by for example disabling data reps when you specify a cast back to the underlying type. So `GET images?id=123&select=raw_data::bytea` where `raw_data` is a `base64` domain with `bytea` storage bypasses the data reps. Without the cast you get ("binary") base64. But then back to the issue that the user might not expect data reps to be possible to bypass. That makes me uncomfortable.

Reply:
> One could also imagine data reps being used as a security thing. Dumb example but imagine a data rep that lets you write your credit card number into a table but when you try to read it back it only gives you the last 4 digits.

This is data masking right? In that case I think this would be better handled by a dedicated pg extension like [postgresql_anonymizer](https://gitlab.com/dalibo/postgresql_anonymizer).

> I'm not sure how this feature should play out when the requested content type (or the payload content type) isn't JSON.

This is one thing that worries me a bit. So we're mostly know for being a JSON API, but REST is much more than that, I'm actively working on https://github.com/PostgREST/postgrest/pull/1582 which would enable us to cover all resource representations(not just JSON). For example we could  output map tiles(see [example](https://github.com/steve-chavez/postgrest-postgis-example#mvt)), HTML, and other binary formats. Personally I'm very excited of the possibilities. 

(Btw, I've just noted there's some conflation here because "representations" is also a REST term, but used more broadly,  more about "rows representations"[XML, JSON, etc) while "data representations" are more about "columns representations")

However, to work on https://github.com/PostgREST/postgrest/pull/1582, I've had to refactor and clear many assumptions made on the code(there's still some left). So what I'd like to ensure here is that the assumption of datareps only working for JSON(`application/json`) is made clear on the code. This way when I finally implement https://github.com/PostgREST/postgrest/pull/1582 I don't have to break any existing functionality.

Btw, do you think "column representations" would be a more apt name?

Reply:
> > One could also imagine data reps being used as a security thing. Dumb example but imagine a data rep that lets you write your credit card number into a table but when you try to read it back it only gives you the last 4 digits.
> 
> This is data masking right? In that case I think this would be better handled by a dedicated pg extension like [postgresql_anonymizer](https://gitlab.com/dalibo/postgresql_anonymizer).
> 

That's a cool extension! 

While the postgresql_anonymizer extension seems like a great tool for data masking, it's unfortunately not on Google Cloud SQL's Postgres extension whitelist. Data reps work have a kind of simplicity to them and I think this type of use case emerge naturally from the idea of input != output format and we want to insert a transformation pipeline.

> > I'm not sure how this feature should play out when the requested content type (or the payload content type) isn't JSON.
> 
> This is one thing that worries me a bit. So we're mostly know for being a JSON API, but REST is much more than that, I'm actively working on #1582 which would enable us to cover all resource representations(not just JSON). For example we could output map tiles(see [example](https://github.com/steve-chavez/postgrest-postgis-example#mvt)), HTML, and other binary formats. Personally I'm very excited of the possibilities.
> 
First a correction, when I wrote I'm not sure how it should play out, I'm not sure what I was worrying about specifically because I *did* think about how this plays out for other things than JSON output.

So I completely agree REST is more than JSON and that's in fact why I wrote the code we're talking about the way I did. Instead of hardcoding that ResolverContext makes JSON, that its outputType is `json`, I made it an argument. So as we keep building on the data reps feature in PostgREST, we can stick other stuff there, like binary or CSV. Then as an end user you can just define more data rep functions which implicitly cast not from domain to json but to `csv` or `html` (which would be text domains) or even `image/png`. So data reps help you get the right representation, like the name implies, even in the presence of multiple potential formats. You want to encode your data differently depending on the kind of output you want, which relates to your #1582 work, if I'm not mistaken.

I think data reps offer powerful compositional capabilities that align well with the relational nature of databases. HTML is a great example where keying your transformation on your types just really lets you compose freely with JOINs and whatever way you get your data into the output, no matter how complex your query is it all just works. Your data will be turned into the format you want at the end because the types follow the values.

Maybe the output is a HTML `<table>`. You want to make cells and right-align your numbers so `5::int` -> `<td class='numeric'>5</td>` when your output type is of the `html_cell` domain. But the same value coming from a `text` column might be `<td>5</td>` to make it left-aligned. I think data reps really give you some fantastic compositional powers in this way. Even if that `5::int` came from some crazy Postgres wizardry, a CTE invoking a procedure making a record set, then joined into your main query... You get the idea. Still works.

Anyhow, the hardcoded `json` output is just a starting point that preserves the status quo while laying out the groundwork for the future.

> (Btw, I've just noted there's some conflation here because "representations" is also a REST term, but it's more about "rows representations"[XML, JSON, etc] while "data representations" is more about "columns representations")
> 
It's both, isn't it? Your columns very likely represent differently if your rows are XML.

> However, to work on #1582, I've had to refactor and clear many assumptions made on the code(there's still some left). So what I'd like to ensure here is that the assumption of datareps only working for JSON(`application/json`) is made clear on the code. This way when I finally implement #1582 I don't have to break any existing functionality.
> 
> Btw, do you think "column representations" would be a more apt name?

I mean there's nothing that says the data is necessarily a column, it could be the output of a calculation or a procedure too. It's somewhat columnar in that its associated to columns in the rows but it's not that clean semantically.

Reply:
> You want to make cells and right-align your numbers so 5::int -> <td class='numeric'>5</td> when your output type is the html_cell domain. The same value coming from a text column might be <td>5</td> to make it left-aligned

That's amazing. I actually been experimenting with HTML inside functions and doing the transformation inside a custom type would be cleaner.

> Now it’s a “table representation”, isn’t it? So we need a broader term.

Maybe we can just leave it at "representations", with different ways to do them. It would encompass both #1582 and #2523.

Reply:
Oh yeah, just "representations" is even more to the point. (Indeed that's the name of my working branch!)